cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -615,36 +609,48 @@ set(Boost_ADDITIONAL_VERSIONS
     "1.60.0"
     "1.60")
 
-if(BOOST_SOURCE STREQUAL "AUTO")
-  find_package(BoostAlt ${ARROW_BOOST_REQUIRED_VERSION})
-  if(NOT BoostAlt_FOUND)
-    build_boost()
-  endif()
-elseif(BOOST_SOURCE STREQUAL "BUNDLED")
-  build_boost()
-elseif(BOOST_SOURCE STREQUAL "SYSTEM")
-  find_package(BoostAlt ${ARROW_BOOST_REQUIRED_VERSION} REQUIRED)
-endif()
-
-if(TARGET Boost::system)
-  set(BOOST_SYSTEM_LIBRARY Boost::system)
-  set(BOOST_FILESYSTEM_LIBRARY Boost::filesystem)
-  set(BOOST_REGEX_LIBRARY Boost::regex)
-elseif(BoostAlt_FOUND)
-  set(BOOST_SYSTEM_LIBRARY ${Boost_SYSTEM_LIBRARY})
-  set(BOOST_FILESYSTEM_LIBRARY ${Boost_FILESYSTEM_LIBRARY})
-  set(BOOST_REGEX_LIBRARY ${Boost_REGEX_LIBRARY})
+if(ARROW_BUILD_INTEGRATION
+   OR ARROW_BUILD_TESTS
```

Comment:
I'm a little worried about how we're going to maintain this `if` statement. Suppose the compute module needs boost for something. How are we going to be sure that `OR ARROW_COMPUTE` gets added here? We won't catch it in our C++ tests because `ARROW_BUILD_TESTS` will cause boost to be included, and we won't catch it in R or Python because `ARROW_PARQUET` is on. But then if someone builds with `ARROW_COMPUTE` but not these other flags, it will fail. 

I don't have a solution in mind, just wanted to express my fear.

Reply:
Yes, I would like to have a more declarative way to define dependencies between components and their build toolchain dependencies rather than doing this ad hoc stuff. The only way to really be sure is to test a battery of different build configurations and make sure they all work